gloss/Graphics/Gloss/Geometry/Triangulation.hs

```diff
@@ -0,0 +1,388 @@
+replaceEndOf :: Vertex -> Edge -> Edge -> Events -> Events
+replaceEndOf p e1 e2 events = let (startOf, endOf) = events Map.! p
+                                  newEndOf = Set.insert e2 $ Set.delete e1 endOf
+                              in Map.insert p  (startOf, newEndOf) events
+
+-- Utility function for safe lookupMin, since set breaks when empty.
+lookupMin :: Set.Set a -> Maybe a
+lookupMin set
+        | Set.null set = Nothing
+        | otherwise = Just $ Set.findMin set
+
+-- Utility function for safe lookupMax, since set breaks when empty.
+lookupMax :: Set.Set a -> Maybe a
+lookupMax set
+        | Set.null set = Nothing
+        | otherwise = Just $ Set.findMax set
+
+-- | Transforms @Path@ into a planar graph with verteces containing all verteces of @Path@ and intersections of path segments.
+polygonGraph :: Path -> Vertices
+
+polygonGraph l = snd $ traverseVertices accumulatorFunction ( Set.empty, Map.empty) $ initialEvents (makeInitialVertexSet l)
+
+-- | Utility function for checking orientation of vertices.
+pointsPositiveOrientation :: Vertex -> Vertex -> Vertex -> Ordering
+pointsPositiveOrientation (Vertex (x0,y0)) (Vertex (x1,y1)) (Vertex (x2,y2)) = compare ((x1-x0)*(y2-y0)) ((x2-x0)*(y1-y0))
+
+-- | Angle between two vectors.
+oriantedAngleVV :: Vector -> Vector -> Float
+
+oriantedAngleVV v1 v2 = normalizeAngle $ argV v1 - argV v2
+
+-- | Checking if verteces are turnign left of right.
+compareAngles :: Vertex -> Vertex -> Vertex -> Vertex -> Ordering
+
+compareAngles (Vertex v1) (Vertex v2) (Vertex v3) (Vertex v4) = compare (oriantedAngleVV (v1-v2) (v3-v2)) (oriantedAngleVV (v1-v2) (v4-v2))
+
+-- | Utility function for adding neighbouring verteces.
+addNeighbour ::  Vertex -> Vertex -> Vertices -> Vertices
+
+addNeighbour point1 point2 verteces = helper point2 point1 $ helper point1 point2 verteces
+                          where helper p1 p2 = Map.insertWith Set.union p1 (Set.singleton p2)
+
+-- | Utility function for removing neighbouring verteces.
+removeNeighbors :: (Vertex, Vertex) -> Vertices -> Vertices
+
+removeNeighbors (point1, point2) verteces = helper point2 point1 $ helper point1 point2 verteces
+                          where helper p1 p2 v = let p1Set = v Map.! p1
+                                                     newP1Set = Set.delete p2 p1Set
+                                                 in if Set.null newP1Set then Map.delete p1 v else Map.insert p1 newP1Set v
+
+-- | Breaks @Path@ into a list of simple paths.
+breakUpToSimplePolygons :: Path -> [Path]
+
+breakUpToSimplePolygons path = traversePolygonGraph Nothing [] (\_ _ _ -> True) Set.empty (polygonGraph path)
+
+-- | Traverse planar graph and calculates simple polygons that bound inner fealds of given graph.
+traversePolygonGraph :: Maybe (Vertex, Vertex) -> [Path] -> (Vertices -> Vertex -> Vertex -> Bool) -> Set.Set (Vertex, Vertex) -> Vertices -> [Path]
```

Comment:
Please add haddock comments that describe what each of these parameters do. For example, what does the deciderFn do?